hare/broker.go

```diff
@@ -104,49 +112,72 @@ func (b *Broker) eventLoop() {
 				b.Warning("Message validation failed: InnerMsg is nil")
 				continue
 			}
-			expInstId := b.maxReg + 1 //  max expect current max + 1
+
 			msgInstId := InstanceId(hareMsg.InnerMsg.InstanceId)
-			// far future unregistered instance
-			if msgInstId > expInstId {
-				b.Warning("Message validation failed: InstanceId. Max: %v Actual: %v", expInstId, msgInstId)
+			state, exist := b.layerState[msgInstId]
+			if exist && state == invalid { // invalid instance, ignore
+				b.Info("%v instance is invalid. Ignoring message", msgInstId)
 				continue
 			}
 
-			// near future
-			if msgInstId == expInstId {
-				futureMsg = true
+			c, ok := b.outbox[msgInstId]
+			if !ok { // unknown instance, maybe just an early msg
+				if b.latestLayer+1 != msgInstId { // not an early msg
+					// ignore msg
+					b.Warning("Message validation failed: InstanceId. Expected: %v Actual: %v", b.latestLayer+1, msgInstId)
```

Comment:
We get here only if the message is not for the current layer, and also not for the next layer, right?
If so, I think that the warning message should print Expected: b.latestLayer, which is the current layer.

Reply:
Yes but this is not what the if checks. It is better to write what was the expected condition that failed.